### PR TITLE
Issue 249/compose support

### PR DESCRIPTION
**Pull request by matzuk**
What remains:

- [x] autoscroll behavior interceptor for Jetpack Compose
- [x] compose (or-then) functionality for Jetpack Compose
- [x] Jetpack Compose tests on the JVM environment
- [x] Readme
- [ ] Add ".idea" to VCS according to [this article](https://proandroiddev.com/deep-dive-into-idea-folder-in-android-studio-53f867cf7b70)

### CODE REVIEW

File: build-conventions/android-convention/src/main/kotlin/convention.android-base.gradle.kts [hunk 1]
```diff
@@ -8,8 +8,25 @@ configure<BaseExtension> {
         named("androidTest").configure { java.srcDir("src/androidTest/kotlin") }
     }
 
-    buildToolsVersion("29.0.3")
-    compileSdkVersion(29)
+    // temp solution of the following bug
```
Comment (matzuk) [inline build-conventions/android-convention/src/main/kotlin/convention.android-base.gradle.kts:11]: @RuslanMingaliev need your help here

File: build-conventions/android-convention/src/main/kotlin/convention.android-library.gradle.kts [hunk 2]
```diff
@@ -11,10 +11,11 @@ android {
         }
     }
 
-    @Suppress("UnstableApiUsage")
-    onVariants {
-        androidTest {
-            enabled = false
-        }
-    }
+    // todo doesn't work on AGP 4.2.2
```
Comment (matzuk) [inline build-conventions/android-convention/src/main/kotlin/convention.android-library.gradle.kts:14]: @RuslanMingaliev such api is not available now

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -1,13 +1,15 @@
 [versions]
-kotlin = "1.4.32"
+kotlin = "1.5.21"
```
Comment (matzuk) [inline gradle/libs.versions.toml:2]: 1.5.31 causes weird problems, postponed

File: build-conventions/android-convention/src/main/kotlin/convention.android-base.gradle.kts [hunk 4]
```diff
@@ -8,8 +8,25 @@ configure<BaseExtension> {
...
+    // https://stackoverflow.com/questions/67358179/android-espresso-test-error-no-static-method-loadsingleserviceornull
+    // in the Kaspresso, it is caused by androidx.compose.ui:ui-test-junit4:1.0.3 in Kakao-compose that uses androidx.test.espresso:espresso-core:3.3.0 under the hood
+    // but we expect using of androidx.test.espresso:espresso-core:3.4.0
+    configurations.all {
+        resolutionStrategy {
+            force("androidx.test:monitor:1.4.0")
+        }
+    }
+
+    // temp solution
```
Comment (matzuk) [inline build-conventions/android-convention/src/main/kotlin/convention.android-base.gradle.kts:21]: @RuslanMingaliev need your help here

File: kaspresso/build.gradle.kts [hunk 5]
```diff
@@ -10,10 +10,12 @@ publish {
 dependencies {
     api(projects.kautomator)
     api(libs.kakao)
+//    api(libs.kakaoCompose)
```
Comment (matzuk) [inline kaspresso/build.gradle.kts:13]: temp until KakaoCompose 0.0.3

File: kaspresso/build.gradle.kts [hunk 6]
```diff
@@ -10,10 +10,12 @@ publish {
...
     api(libs.bundles.espresso)
     api(libs.uiAutomator)
     api(libs.androidXCore)
     api(libs.androidXRules)
+    api(files("libs/compose-debug.aar"))
```
Comment (matzuk) [inline kaspresso/build.gradle.kts:18]: temp until KakaoCompose 0.0.3

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt [hunk 7]
```diff
@@ -47,32 +47,16 @@ class ObjectAutoScrollProviderImpl(
...
-         * Scrolls to the bottom and looks for the given view. Invokes the action if the view was found.
-         */
-        do {
-            if (interaction.tryToFindUiObject()) {
-                logger.i("UiView autoscroll successfully performed.")
-                return action.invoke()
-            }
-        } while (scrollable.scrollForward())
-
-        /**
```
Comment (matzuk) [inline kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt:65]: Removed scrolling up. I think that such behavior is a little bit tricky and unpredictable. That's why the only scrolling bottom remains.

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behaviorkautomator/impl/flakysafety/FlakySafeObjectBehaviorInterceptor.kt [hunk 8]
```diff
@@ -30,7 +30,10 @@ class FlakySafeObjectBehaviorInterceptor(
         interaction: UiObjectInteraction,
         assertion: UiObjectAssertion,
         activity: () -> T
-    ): T = flakySafely(action = activity)
+    ): T = flakySafely(action = {
+        interaction.reFindUiObject()
```
Comment (matzuk) [inline kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behaviorkautomator/impl/flakysafety/FlakySafeObjectBehaviorInterceptor.kt:34]: Improvements of Kautomator stability

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/kaspresso/Kaspresso.kt [hunk 9]
```diff
@@ -819,10 +857,12 @@ data class Kaspresso(
                         instrumentalDependencyProviderFactory.getInterceptorProvider<SystemDialogSafetyViewBehaviorInterceptor>(instrumentation),
                         adbServer
                     ),
-                    FlakySafeViewBehaviorInterceptor(flakySafetyParams, libLogger)
+                    FlakySafeViewBehaviorInterceptor(flakySafetyParams, libLogger),
+                    FailureLoggingViewBehaviorInterceptor(libLogger)
```
Comment (matzuk) [inline kaspresso/src/main/kotlin/com/kaspersky/kaspresso/kaspresso/Kaspresso.kt:861]: Why have we forgotten about all FailureLogging***Interceptors???

File: gradle/libs.versions.toml [hunk 10]
```diff
@@ -1,13 +1,15 @@
...
+espresso = "3.4.0"
+allure = "2.2.7"
+compose = "1.0.2"
+activityCompose = "1.3.1"
 
 [libraries]
 # plugins
 kotlinPlugin = { module = "org.jetbrains.kotlin:kotlin-gradle-plugin", version.ref = "kotlin" }
-androidPlugin = "com.android.tools.build:gradle:4.1.0"
+androidPlugin = "com.android.tools.build:gradle:4.2.2"
```
Comment (eakurnikov) [inline gradle/libs.versions.toml:12]: Мб нам его до 7 поднять?

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/tolibrary/compose/ComposeSemanticsInterceptor.kt [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    override fun interceptPerform(interaction: ComposeInteraction, action: ComposeAction) {
+        composeSemanticsBehaviorInterceptors.fold(
+            initial = {
+                composeSemanticsWatcherInterceptors.forEach {
+                    it.interceptPerform(interaction, action)
+                }
+                interaction.perform(action)
+            },
+            operation = {
+                    acc, deviceBehaviorInterceptor ->
```
Comment (eakurnikov) [inline kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/tolibrary/compose/ComposeSemanticsInterceptor.kt:52]: Линт внутри меня говорит, что тут форматирование поехало)0)

File: kautomator/src/main/kotlin/com/kaspersky/components/kautomator/intercept/interaction/UiObjectInteraction.kt [hunk 12]
```diff
@@ -16,20 +16,15 @@ class UiObjectInteraction(
...
-    var uiObject2: UiObject2? = null
+    var uiObject2: UiObject2? = calculateUiObject()
         private set
 
-    /**
-     * Tries to find UiObject2 with given selector
-     * @return true if object was found, false otherwise
-     */
-    fun tryToFindUiObject(): Boolean {
+    private fun calculateUiObject(): UiObject2? {
```
Comment (eakurnikov) [inline kautomator/src/main/kotlin/com/kaspersky/components/kautomator/intercept/interaction/UiObjectInteraction.kt:22]: А почему calculate то? Она вроде не считает ничего

File: samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/resources/C.kt [hunk 13]
```diff
@@ -0,0 +1,19 @@
+package com.kaspersky.kaspressample.jetpackcompose.resources
+
+// Like R, but C
+object C {
```
Comment (eakurnikov) [inline samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/resources/C.kt:4]: Это твое изобретение или это паттерн типо рекомендованный? Элементы по строке ищутся?

**Review by eakurnikov — DISMISSED** [review]
Nnoooice

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt (hunk 7 above)
Comment (RuslanMingaliev) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt:65]: It is a bad idea to change the scroll behavior. You can break users' tests

Comment (sergio-sastre) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt:65]: Moreover, it could be that the user starts the screen in the middle of the list for whatever reason, like a highlighted element or something similar. I believe it is better to keep it as well

File: kautomator/src/main/kotlin/com/kaspersky/components/kautomator/component/common/assertions/UiBaseAssertions.kt [hunk 14]
```diff
@@ -26,7 +27,18 @@ interface UiBaseAssertions {
      * Checks if the view is not completely displayed
      */
     fun isNotDisplayed() {
-        assertThat(view.loadView()).isFalse()
+        // the implementation is a little bit tricky because
+        // UiObjectInteractionDelegate throws UnfoundedUiObjectException in case of absence of a related UIObject2
+        // that's why we just catch the mentioned exception and do a final check
+        val notDisplayed = try {
```
Comment (RuslanMingaliev) [inline kautomator/src/main/kotlin/com/kaspersky/components/kautomator/component/common/assertions/UiBaseAssertions.kt:33]: Why did you change the previous behavior?

File: samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/features/flaky/SimpleFlakyViewModel.kt [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    private val _mutableSimpleFlakyState = MutableLiveData(
+        SimpleFlakyState(
+            firstButtonVisibility = true,
+            secondButtonVisibility = false,
+            editTextVisibility = false,
+            editText = "Some text"
+        )
+    )
+
+    val simpleFlakyStateLiveData: LiveData<SimpleFlakyState> = _mutableSimpleFlakyState
```
Comment (RuslanMingaliev) [inline samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/features/flaky/SimpleFlakyViewModel.kt:28]: the backing property must have the same name starting with an underscore
`_simpleFlakyStateLiveData`

File: gradle/libs.versions.toml [hunk 16]
```diff
@@ -17,28 +19,47 @@ detektFormatting = { module = "io.gitlab.arturbosch.detekt:detekt-formatting", v
 kotlinStdlib = { module = "org.jetbrains.kotlin:kotlin-stdlib", version.ref = "kotlin" }
 espressoCore = { module = "androidx.test.espresso:espresso-core", version.ref = "espresso" }
 espressoWeb = { module = "androidx.test.espresso:espresso-web", version.ref = "espresso" }
-testCore = "androidx.test:core:1.3.0"
+testCore = "androidx.test:core:1.4.0"
 uiAutomator = "androidx.test.uiautomator:uiautomator:2.2.0"
-fragmentTesting = "androidx.fragment:fragment-testing:1.3.2"
+fragmentTesting = "androidx.fragment:fragment-testing:1.3.6"
+# upgrading to 4.6.1 is blocked by https://github.com/robolectric/robolectric/issues/6593
```
Comment (RuslanMingaliev) [inline gradle/libs.versions.toml:25]: It's not a bug, there is a solution in the comments

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt (hunk 7 above)
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/ObjectAutoScrollProviderImpl.kt:65]: Agree. I just remembered a case when up scrolling caused the browser page reloading that was inappropriate. But, yep, you are right.

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behaviorkautomator/impl/flakysafety/FlakySafeObjectBehaviorInterceptor.kt (hunk 8 above)
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/behaviorkautomator/impl/flakysafety/FlakySafeObjectBehaviorInterceptor.kt:34]: 1. This one must be relocated to a separate Interceptor. Thanks that you've highlighted.
2. Good question about stability. So, the concept is similar to the one that we had in Loader...Interceptor. But, the architecture solution is better and the same with Jetpack Compose Intercepting (see old weird methods like `loadView` and etc.). Better stability is achieved due to changing the order of behavior interceptors (try to reload earlier). I think, there were cases when SystemDialog and Scrolling Interceptors consumed all allocated time, that's why there was no time to reload Element. But, I don't have tests where I can check it.

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/kaspresso/Kaspresso.kt (hunk 9 above)
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/kaspresso/Kaspresso.kt:864]: No, we need. These interceptors write good exceptions in Logs including nested or compose exceptions.

File: gradle/libs.versions.toml (hunk 10 above)
Comment (matzuk) [reply gradle/libs.versions.toml:12]: I tried but I got a lot of errors with our plugins =(

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/tolibrary/compose/ComposeSemanticsInterceptor.kt (hunk 11 above)
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/interceptors/tolibrary/compose/ComposeSemanticsInterceptor.kt:52]: Will check it

File: kautomator/src/main/kotlin/com/kaspersky/components/kautomator/intercept/interaction/UiObjectInteraction.kt (hunk 12 above)
Comment (matzuk) [reply kautomator/src/main/kotlin/com/kaspersky/components/kautomator/intercept/interaction/UiObjectInteraction.kt:22]: Yep, maybe there is a sense to put another name.

File: samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/resources/C.kt (hunk 13 above)
Comment (matzuk) [reply samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/resources/C.kt:4]: @RuslanMingaliev suggested such an approach. I like it =)

File: kaspresso/build.gradle.kts [hunk 17]
```diff
@@ -18,6 +19,7 @@ dependencies {
     implementation(libs.kotlinStdlib)
     implementation(libs.gson)
     implementation(projects.adbServer.adbserverDevice)
+    implementation(libs.composeUiTest)
```
[the remark this answers is not in the document]
Comment (matzuk) [reply kaspresso/build.gradle.kts:22]: Yep!!

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+     *
+     * @return the result of [action] invocation.
+     */
+    override fun <ActionType> passAction(
+        elementLoader: () -> Unit,
+        action: () -> ActionType
+    ): ActionType {
+        return try {
+            action.invoke()
+        } catch (error: Throwable) {
```
[the remark this answers is not in the document]
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt:27]: Very strange comment, honestly. Please, describe your suggestion better =)

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+            logger.i("Reloading of the element is started")
+            elementLoader.invoke()
+            logger.i("Reloading of the element is finished")
+            logger.i("Repeat action again with the reloaded element")
```
[the remark this answers is not in the document]
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt:31]: No, it's not. Have a look at logs running examples.

File: kautomator/src/main/kotlin/com/kaspersky/components/kautomator/component/common/assertions/UiBaseAssertions.kt (hunk 14 above)
Comment (matzuk) [reply kautomator/src/main/kotlin/com/kaspersky/components/kautomator/component/common/assertions/UiBaseAssertions.kt:33]: The previous solution required an additional method of `UiObjectInteractionDelegate` which was too poor because made an abstraction leak.

File: samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/features/flaky/SimpleFlakyViewModel.kt (hunk 15 above)
Comment (matzuk) [reply samples/kaspresso-sample/src/main/kotlin/com/kaspersky/kaspressample/jetpackcompose/features/flaky/SimpleFlakyViewModel.kt:28]: Agree

File: gradle/libs.versions.toml (hunk 16 above)
Comment (matzuk) [reply gradle/libs.versions.toml:25]: I am not sure that you are right. Ok, can you try to fix it?

File: kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt (hunk 18 above)
[the remark this answers is not in the document]
Comment (RuslanMingaliev) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt:27]: It's a very bad practice to catch a too generic exception. Do you know the types of exceptions that could be thrown here?

[the remark this answers is not in the document]
Comment (matzuk) [reply kaspresso/src/main/kotlin/com/kaspersky/kaspresso/elementloader/ElementLoaderProviderImpl.kt:27]: Got it. I think the answer is "yes".

**Review by RuslanMingaliev — APPROVED** [review]
